spv/sync.go

```diff
@@ -0,0 +1,1428 @@
+	// needing to keep all relevant data in memory.
+	rescanFilter map[int]*wallet.RescanFilter
+	filterData   map[int]*blockcf.Entries
+	filterMu     sync.Mutex
+
+	// seenTxs records hashes of received inventoried transactions.  Once a
+	// transaction is fetched and processed from one peer, the hash is added to
+	// this cache to avoid fetching it again from other peers that announce the
+	// transaction.
+	seenTxs lru.Cache
+
+	// Sidechain management
+	sidechains  wallet.SidechainForest
+	sidechainMu sync.Mutex
+
+	currentLocators   []*chainhash.Hash
+	locatorGeneration uint
+	locatorMu         sync.Mutex
+
+	// Holds all potential callbacks used to notify clients
+	notifications *Notifications
+}
+
+// Notifications struct to contain all of the upcoming callbacks that will
+// be used to update the rpc streams for syncing.
+type Notifications struct {
+	Synced                       func(walletID int, sync bool)
+	PeerConnected                func(peerCount int32, addr string)
+	PeerDisconnected             func(peerCount int32, addr string)
+	FetchMissingCFiltersStarted  func(walletID int)
+	FetchMissingCFiltersProgress func(walletID int, startCFiltersHeight, endCFiltersHeight int32)
+	FetchMissingCFiltersFinished func(walletID int)
+	FetchHeadersStarted          func(peerInitialHeight int32)
+	FetchHeadersProgress         func(lastHeaderHeight int32, lastHeaderTime int64)
+	FetchHeadersFinished         func()
+	DiscoverAddressesStarted     func(walletID int)
+	DiscoverAddressesFinished    func(walletID int)
+	RescanStarted                func(walletID int)
+	RescanProgress               func(walletID int, rescannedThrough int32)
+	RescanFinished               func(walletID int)
+
+	// MempoolTxs is called whenever new relevant unmined transactions are
+	// observed and saved.
+	MempoolTxs func(walletID int, txs []*wire.MsgTx)
+
+	// TipChanged is called when the main chain tip block changes.
+	// When reorgDepth is zero, the new block is a direct child of the previous tip.
+	// If non-zero, one or more blocks described by the parameter were removed from
+	// the previous main chain.
+	// txs contains all relevant transactions mined in each attached block in
+	// unspecified order.
+	// reorgDepth is guaranteed to be non-negative.
+	TipChanged func(tip *wire.BlockHeader, reorgDepth int32, txs []*wire.MsgTx)
+}
+
+// NewSyncer creates a Syncer that will sync the wallet using SPV.
+func NewSyncer(wallets map[int]*wallet.Wallet, lp *p2p.LocalPeer) *Syncer {
+	rescanFilter := make(map[int]*wallet.RescanFilter)
+	filterData := make(map[int]*blockcf.Entries)
+	atomicCatchUpTryLocks := make(map[int]*uint32, len(wallets))
+	atomicWalletsSynced := make(map[int]*uint32, len(wallets))
+
+	for walletID := range wallets {
+		rescanFilter[walletID] = wallet.NewRescanFilter(nil, nil)
+		filterData[walletID] = &blockcf.Entries{}
+		atomicCatchUpTryLocks[walletID] = new(uint32)
+		atomicWalletsSynced[walletID] = new(uint32)
```

Comment:
Are these necessary? I remember a previous review comment recommending leaving these out as the slice gets created with a default value of 0 per index.

Reply:
Removing it causes a nil pointer exception.